File: dsa/9.stack_and_queue/2696_Minimum_String_Length_After_Removing_Substrings.py
import typing as t

class Solution:
    def minLength(self, s: str) -> int:
        char_dict: t.Dict[str, str] = {
            "B": "A"
            , "D": "C"
        }
        stack: t.List[str] = []

        for char in s:
            if stack and char in char_dict.keys() and stack[-1] == char_dict.get(char):
                stack.pop()
            else:
                stack.append(char)
        
        return len(stack)

File: dsa/9.stack_and_queue/2696_Minimum_String_Length_After_Removing_Substrings.java
package stack_and_queue;
import java.util.*;

class Solution {
    public int minLength(String s) {
        Map<Character, Character> map = new HashMap<Character, Character>();
        map.put('D', 'C');
        map.put('B', 'A');
        Stack<Character> charStack = new Stack<Character>();
        int len = s.length();

        for(int i = 0; i < s.length(); i++){
            if (charStack.isEmpty() == false 
                && map.containsKey(s.charAt(i)) == true
                && map.get(s.charAt(i)) == charStack.peek()){
                    charStack.pop();
                    len -= 2;
            }
            else{
                charStack.push(s.charAt(i));
            }
        }
        return len;
    }
}
